### PR TITLE
[ism8] Add UoM support

### PR DESCRIPTION
**Pull request by lsiepel**
Signed-off-by: lsiepel <leosiepel@gmail.com>

- Adds UoM to all applicable channels.
- Adds several units tests for in- and outgoing messages.
- Minor refactoring of Thing handler.

Due to the change in channels, this is not backwards compatible. Channel configuration and linked Items have to be adapted.

Test jar for 3.4.1 is available here: https://1drv.ms/u/s!AnMcxmvEeupwjp4L347Q22cnMI1QQg?e=ROo8Y1
Tests have been performed by @schmidmuc (thanks for that!)

Fixes: https://github.com/openhab/openhab-addons/issues/9598

Also replaces: https://github.com/openhab/openhab-addons/pull/14665

### CODE REVIEW

**Comment by holgerfriedrich** [comment]
As this is a breaking change, did you think of adding a note to update.lst in distro repo?
https://github.com/openhab/openhab-distro/blob/main/distributions/openhab/src/main/resources/bin/update.lst

**Comment by lsiepel** [comment]
> As this is a breaking change, did you think of adding a note to update.lst in distro repo? https://github.com/openhab/openhab-distro/blob/main/distributions/openhab/src/main/resources/bin/update.lst

Yeah, thought about it, never done that before, hope it is right.

**Review by Confectrician — COMMENTED** [review]
Looks good for the docs part.

**Comment by lsiepel** [comment]
Wanted to add the ` instructions.xml` update instructions, but i fail to find any documentation how to do this. @jlaur can you share some links/hints ?

**Comment by jlaur** [comment]
> Wanted to add the ` instructions.xml` update instructions, but i fail to find any documentation how to do this. @jlaur can you share some links/hints ?

There is no documentation: https://github.com/openhab/openhab-core/pull/3330#issuecomment-1432033039

A few sources for information:
- openhab/openhab-core#3330 - see unit tests.
- #14572
- #14473
- https://github.com/openhab/openhab-addons/pull/14583#discussion_r1134688331

**Comment by jlaur** [comment]
> > Wanted to add the ` instructions.xml` update instructions, but i fail to find any documentation how to do this. @jlaur can you share some links/hints ?
> 
> There is no documentation: [openhab/openhab-core#3330 (comment)](https://github.com/openhab/openhab-core/pull/3330#issuecomment-1432033039)

Now there is: https://next.openhab.org/docs/developer/bindings/thing-xml.html#updating-thing-types

**Comment by lsiepel** [comment]
> a

Oke, i have read some more and used it elsewhere. So i can understand the basics, but now this binding is different, it adds the channels during setup dynamically.

I have this problem with the following scenario:
For example old channel-type `number-readonly`  is split into multiple channel-types `percentage-r`, `temperature-r` and `flowrate-r`. Maybe i can make some rules to determine those rules, but i don't think the instructions.xml features some conditional logic, does it?

If not, this PR is ready to review/merge and the breaking change PR will still be needed.

**Comment by lsiepel** [comment]
@openhab/add-ons-maintainers This PR has been tested and confirmed to be working as expected. Can we get this merged?

**Comment by lsiepel** [comment]
Seen some other PR's and for those cases when channels where automatically added the thing upgrade was not needed. So i think this is all set now.

**Review by jlaur — CHANGES_REQUESTED** [review]
Not a review, sorry, but I noticed two old comments pending, so thought I'd "flush" them. 🙂

File: bundles/org.openhab.binding.ism8/README.md [hunk 1]
```diff
@@ -34,15 +34,9 @@ The ISM8 does currently support 4 different devices at the same moment of time (
...
-| Number | DpId004 | "Kesseltemperatur"         | id, type, write |
-
-Type:
-
-- Switch use for boolean values
-- Number use for any number
-- Other types may work as well.
+| Type   | Name    | Description        | Configuration |
+|--------|---------|--------------------|---------------|
+| Number | DpId004 | "Kesseltemperatur" |               |
```
[the remark this answers is not in the document]
Comment (lsiepel) [reply bundles/org.openhab.binding.ism8/README.md:39]: true

File: bundles/org.openhab.binding.ism8/src/main/java/org/openhab/binding/ism8/internal/util/Ism8DomainMap.java [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+import org.openhab.core.library.unit.SIUnits;
+import org.openhab.core.library.unit.Units;
+import org.openhab.core.types.Command;
+import org.openhab.core.types.State;
+import org.openhab.core.types.UnDefType;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * The {@link DataPointBase} is the base class for all data points
```
[the remark this answers is not in the document]
Comment (lsiepel) [reply bundles/org.openhab.binding.ism8/src/main/java/org/openhab/binding/ism8/internal/util/Ism8DomainMap.java:35]: Copy/paste error.

**Comment by lsiepel** [comment]
@openhab/add-ons-maintainers It would be nice to have this available in 4.2 M1 Anyone able to review ?

**Review by kaikreuzer — CHANGES_REQUESTED** [review]
Thank you - just some minor comments.

File: bundles/org.openhab.binding.ism8/src/main/java/org/openhab/binding/ism8/internal/Ism8BindingConstants.java [hunk 3]
```diff
@@ -41,4 +41,13 @@ public class Ism8BindingConstants {
      *
      */
     public static final String PORT_NUMBER = "portNumber";
+
+    // Channel Configuration parameters
+
+    /**
+     * Channel id configuration parameter
+     *
+     */
```
[the remark this answers is not in the document]
Comment (lsiepel) [reply bundles/org.openhab.binding.ism8/src/main/java/org/openhab/binding/ism8/internal/Ism8BindingConstants.java:50]: Done

File: bundles/org.openhab.binding.ism8/src/main/java/org/openhab/binding/ism8/internal/Ism8Handler.java [hunk 4]
```diff
@@ -134,25 +114,91 @@ public void connectionStatusChanged(ThingStatus status) {
...
+        if (config.containsKey(CHANNEL_CONFIG_ID) && config.containsKey(CHANNEL_CONFIG_TYPE)) {
+            try {
+                int id = Integer.parseInt(config.get(CHANNEL_CONFIG_ID).toString());
+                String type = config.get(CHANNEL_CONFIG_TYPE).toString();
+                if (svr != null && type != null && description != null) {
+                    svr.addDataPoint(id, type, description);
+                    return true;
+                }
+            } catch (NumberFormatException e) {
+                this.logger.warn(
```
[the remark this answers is not in the document]
Comment (lsiepel) [reply bundles/org.openhab.binding.ism8/src/main/java/org/openhab/binding/ism8/internal/Ism8Handler.java:128]: Done

File: bundles/org.openhab.binding.ism8/src/main/java/org/openhab/binding/ism8/internal/Ism8Handler.java [hunk 5]
```diff
@@ -134,25 +114,91 @@ public void connectionStatusChanged(ThingStatus status) {
...
+                        "Ism8: ID couldn't be converted correctly. Check the configuration of channel {}. Cfg={}",
+                        description, config);
+            }
+        } else {
+            this.logger.debug("Ism8: ID or type missing - Channel={}  Cfg={}", description, config);
```
[the remark this answers is not in the document]
Comment (lsiepel) [reply bundles/org.openhab.binding.ism8/src/main/java/org/openhab/binding/ism8/internal/Ism8Handler.java:133]: Done

File: bundles/org.openhab.binding.ism8/src/main/java/org/openhab/binding/ism8/internal/Ism8Handler.java [hunk 6]
```diff
@@ -134,25 +114,91 @@ public void connectionStatusChanged(ThingStatus status) {
...
+        return false;
+    }
+
+    private void setDataPoint(IDataPoint dataPoint, Command command) {
+        Server svr = this.server;
+        if (svr != null) {
+            try {
+                svr.sendData(Ism8DomainMap.toISM8WriteData(dataPoint, command));
+            } catch (IOException e) {
+                this.logger.debug("Writting to Ism8 DataPoint '{}' failed. '{}'", dataPoint.getId(), e.getMessage());
```
[the remark this answers is not in the document]
Comment (lsiepel) [reply bundles/org.openhab.binding.ism8/src/main/java/org/openhab/binding/ism8/internal/Ism8Handler.java:144]: Done

File: bundles/org.openhab.binding.ism8/src/main/java/org/openhab/binding/ism8/internal/Ism8Handler.java [hunk 7]
```diff
@@ -134,25 +114,91 @@ public void connectionStatusChanged(ThingStatus status) {
...
+        if (config.containsKey(CHANNEL_CONFIG_ID) && config.containsKey(CHANNEL_CONFIG_TYPE)) {
+            try {
+                int id = Integer.parseInt(config.get(CHANNEL_CONFIG_ID).toString());
+                dataPoint = svr.getDataPoint(id);
+            } catch (NumberFormatException e) {
+                this.logger.debug("Retrieving Ism8 DataPoint '{}' failed. '{}'", channel.getConfiguration(),
+                        e.getMessage());
+            }
+        } else {
+            this.logger.debug("Ism8: ID or type missing - Channel={}  Cfg={}", channel.getLabel(),
```
[the remark this answers is not in the document]
Comment (lsiepel) [reply bundles/org.openhab.binding.ism8/src/main/java/org/openhab/binding/ism8/internal/Ism8Handler.java:165]: Done

File: bundles/org.openhab.binding.ism8/src/main/java/org/openhab/binding/ism8/internal/Ism8Handler.java [hunk 8]
```diff
@@ -134,25 +114,91 @@ public void connectionStatusChanged(ThingStatus status) {
...
+            int id = Integer.parseInt(channel.getConfiguration().get(CHANNEL_CONFIG_ID).toString());
+            if (id == dataPoint.getId()) {
+                if (dataPoint.getValueObject() != null) {
+                    this.logger.debug("Ism8: updating channel {} with datapoint: {}", channel.getUID().getAsString(),
+                            dataPoint.getId());
+                    updateState(channel.getUID(), Ism8DomainMap.toOpenHABState(dataPoint));
+                    return true;
+                }
+            } else {
+                this.logger.debug("Ism8 channel: {} and DataPoint do not have a matching Id: {} vs {}",
```
[the remark this answers is not in the document]
Comment (lsiepel) [reply bundles/org.openhab.binding.ism8/src/main/java/org/openhab/binding/ism8/internal/Ism8Handler.java:182]: Done

File: bundles/org.openhab.binding.ism8/src/main/java/org/openhab/binding/ism8/internal/Ism8Handler.java [hunk 9]
```diff
@@ -134,25 +114,91 @@ public void connectionStatusChanged(ThingStatus status) {
...
-                    this.logger.warn(
-                            "Ism8 updateDataPoint: ID couldn't be converted correctly. Check the configuration of channel {}. {}",
-                            channel.getLabel(), e.getMessage());
+            if (channel.getConfiguration().containsKey(CHANNEL_CONFIG_ID)) {
+                if (updateChannel(channel, dataPoint)) {
+                    break;
                 }
             }
         }
+        this.logger.debug("Ism8: no channel was found for DataPoint id: {}", dataPoint.getId());
```
[the remark this answers is not in the document]
Comment (lsiepel) [reply bundles/org.openhab.binding.ism8/src/main/java/org/openhab/binding/ism8/internal/Ism8Handler.java:202]: Done

File: bundles/org.openhab.binding.ism8/src/main/java/org/openhab/binding/ism8/internal/Ism8Handler.java [hunk 10]
```diff
@@ -134,25 +114,91 @@ public void connectionStatusChanged(ThingStatus status) {
...
+                            dataPoint.getId());
+                    updateState(channel.getUID(), Ism8DomainMap.toOpenHABState(dataPoint));
+                    return true;
+                }
+            } else {
+                this.logger.debug("Ism8 channel: {} and DataPoint do not have a matching Id: {} vs {}",
+                        channel.getUID(), id, dataPoint.getId());
+            }
+        } catch (NumberFormatException e) {
+            this.logger.warn(
```
[the remark this answers is not in the document]
Comment (lsiepel) [reply bundles/org.openhab.binding.ism8/src/main/java/org/openhab/binding/ism8/internal/Ism8Handler.java:186]: Done

File: bundles/org.openhab.binding.ism8/src/main/java/org/openhab/binding/ism8/internal/Ism8Handler.java [hunk 11]
```diff
@@ -134,25 +114,91 @@ public void connectionStatusChanged(ThingStatus status) {
...
+        }
+        return dataPoint;
+    }
+
+    private boolean updateChannel(Channel channel, IDataPoint dataPoint) {
+        try {
+            int id = Integer.parseInt(channel.getConfiguration().get(CHANNEL_CONFIG_ID).toString());
+            if (id == dataPoint.getId()) {
+                if (dataPoint.getValueObject() != null) {
+                    this.logger.debug("Ism8: updating channel {} with datapoint: {}", channel.getUID().getAsString(),
```
[the remark this answers is not in the document]
Comment (lsiepel) [reply bundles/org.openhab.binding.ism8/src/main/java/org/openhab/binding/ism8/internal/Ism8Handler.java:176]: Done

File: bundles/org.openhab.binding.ism8/src/main/java/org/openhab/binding/ism8/internal/Ism8Handler.java [hunk 12]
```diff
@@ -47,75 +51,51 @@ public Ism8Handler(Thing thing) {
...
 
         this.updateStatus(ThingStatus.UNKNOWN);
         svr.addDataPointChangeListener(this);
         scheduler.execute(svr::start);
-        this.server = svr;
+    }
+
+    @Override
+    public void handleCommand(ChannelUID channelUID, Command command) {
+        this.logger.debug("Ism8: Handle command = {} {}", channelUID.getId(), command);
```
[the remark this answers is not in the document]
Comment (lsiepel) [reply bundles/org.openhab.binding.ism8/src/main/java/org/openhab/binding/ism8/internal/Ism8Handler.java:77]: Done

**Review by kaikreuzer — APPROVED** [review]
Thanks!

**Pull request merged**